(PE-35382) Add --all action

nmburgan opened the pull request:
This deletes all certs in the signed directory, and must be used by itself.

Additionally, this tweaks the behavior of puppetserver ca delete --expired to not print an error when a cert listed in the inventory is not found on disk, since this will happen when you run the command multiple times and we don't really want people messing with inventory.txt manually to get rid of this error.

shaigy commented on lib/puppetserver/ca/action/delete.rb:
```diff
@@ -192,6 +198,12 @@ def run(args)
           end
 
+          if args['all']
+            certnames = Dir.glob("#{cadir}/signed/*.pem").map{ |c| File.basename(c, '.pem') }
```
This will walk the directory twice, once for getting filenames, and then in delete_certs to delete the files. Probably not too expensive.

nmburgan replied:
I don't think it's actually walking the directory inside `delete_certs`.  `delete_certs` assumes the file is there and constructs a file path from the certname.  However, you're right in that it might be slightly more efficient to have the function delete the file as it walks the directory.  But I think it's pretty negligible.

shaigy commented on lib/puppetserver/ca/action/delete.rb:
```diff
@@ -192,6 +198,12 @@ def run(args)
+          if args['all']
+            certnames = Dir.glob("#{cadir}/signed/*.pem").map{ |c| File.basename(c, '.pem') }
+            # Since we don't run this with any other flags, we can set these variables directly
+            deleted_count, errored = delete_certs(cadir, certnames)
```
Should error_on_not_found be set to false here? Its just walking the directory and deleting them, so I am thinking that error message may be misleading. What do you think?

nmburgan replied:
It theoretically shouldn't ever pass in a certname it can't find, since it's just picking up files on disk, so I don't think the error will ever actually show in the first place.

rileynewton commented:
FYI, has nothing to do with this PR but `delete` is listed under the wrong action type in the cli doc, should be admin action  if you want to update it real quick here.

```
[root@wan-naturalism ca]# puppetserver ca
Unknown action: 
Usage: puppetserver ca <action> [options]

Manage the Private Key Infrastructure for
Puppet Server's built-in Certificate Authority

Available Actions:

  Certificate Actions (requires a running Puppet Server):

    clean	Revoke cert(s) and remove related files from CA
    delete	Delete certificate(s)
    generate	Generate a new certificate signed by the CA
    list	List certificates and CSRs
    revoke	Revoke certificate(s)
    sign	Sign certificate request(s)

  Administrative Actions (requires Puppet Server to be stopped):

    import	Import an external CA chain and generate server PKI
    setup	Setup a self-signed CA chain for Puppet Server
    enable	Setup infrastructure CRL based on a node inventory.
    migrate	Migrate the existing CA directory to /etc/puppetlabs/puppetserver/ca
    prune	Prune the local CRL on disk to remove certificate entries

```

rileynewton commented:
Also it looks like the "(default)" part of the line below `--remove-expired             Remove expired  entries from CRL(default)` should be removed if the description above that is accurate.

```
[root@wan-naturalism ~]# puppetserver ca prune --help
Usage:
  puppetserver ca prune [--help]
  puppetserver ca prune [--config]
  puppetserver ca prune [--config] [--remove-duplicates]
  puppetserver ca prune [--config] [--remove-expired]
  puppetserver ca prune [--config] [--remove-entries] [--serial NUMBER[,NUMBER]] [--certname NAME[,NAME]]

Description:
  Prune the list of revoked certificates. If no options are provided or
  --remove-duplicates is specified, prune CRL of any duplicate entries.
  If --remove-expired is specified, remove expired entries from CRL.
  If --remove-entries is specified, remove matching entries provided by
  --serial and/or --certname values. This command will only prune the CRL
  issued by Puppet's CA cert.

Options:
        --help                       Display this command-specific help output
        --config CONF                Path to the puppet.conf file on disk
        --remove-duplicates          Remove duplicate entries from CRL(default)
        --remove-expired             Remove expired  entries from CRL(default)
        --remove-entries             Remove entries from CRL
        --serial NUMBER[,NUMBER]     Serial numbers(s) in HEX to be removed from CRL
        --certname NAME[,NAME]       Name(s) of the cert(s) to be removed from CRL
```

nmburgan commented:
> FYI, has nothing to do with this PR but `delete` is listed under the wrong action type in the cli doc, should be admin action if you want to update it real quick here.
> 
> ```
> [root@wan-naturalism ca]# puppetserver ca
> Unknown action: 
> Usage: puppetserver ca <action> [options]
> 
> Manage the Private Key Infrastructure for
> Puppet Server's built-in Certificate Authority
> 
> Available Actions:
> 
>   Certificate Actions (requires a running Puppet Server):
> 
>     clean	Revoke cert(s) and remove related files from CA
>     delete	Delete certificate(s)
>     generate	Generate a new certificate signed by the CA
>     list	List certificates and CSRs
>     revoke	Revoke certificate(s)
>     sign	Sign certificate request(s)
> 
>   Administrative Actions (requires Puppet Server to be stopped):
> 
>     import	Import an external CA chain and generate server PKI
>     setup	Setup a self-signed CA chain for Puppet Server
>     enable	Setup infrastructure CRL based on a node inventory.
>     migrate	Migrate the existing CA directory to /etc/puppetlabs/puppetserver/ca
>     prune	Prune the local CRL on disk to remove certificate entries
> ```

Good catch! Fixed, and updated the summary text a bit.
